### DISCUSSION TITLE
creating unique id's

### DISCUSSION THREAD

**Discussion by zerenxyz** [ComfortablyCoding/strapi-plugin-slugify#47] [Q&A]
Is it possible to use slugify and create a UID that takes the attached fields from multiple relations and create a unique ID and such?

I want to have a unique id for each item that I make in my entry but to take for example the first 2 characters from category, 2 first from sub-category, 2 first from item name and add a 3-4 digit

![image](https://user-images.githubusercontent.com/101885283/159132051-65b8a0fe-d81a-4118-9931-5fd367ceb241.png)

or is this something that can be added as a feature to slugify? I am probably able to create this later on in the frontend but would love if it is possible to do it right away as I am making new entries.

_sorry for the duplicate post on my work account didnt realize it haha_

**Comment by ComfortablyCoding** [comment]
At the moment it is possible to create a slug from multiple reference fields by providing an array (introduced in #22 ). I do not believe the plugin accepts relation fields for the reference field option. I will need to investigate this further.

It does not support anything but the full text for them though. I am thinking to add in the option to provide your own transformation function, this way the plugin itself would not need to cover every possible combination use case. What do you think?

> sorry for the duplicate post on my work account didnt realize it haha

No worries, I deleted the duplicate.

**Reply by zerenxyz** [reply]
Heya! Ah I see, would adding user own transformation function be with regex or such? I never used plugins currently from marketplace with Strapi so I am actually unsure how the setup/or how it works, I was hoping it would be possible to just generate it itself like how the UID currently is done within Strapi before you save/publish

**Reply by ComfortablyCoding** [reply]
Just generating the composite UID is already possible via an array is already possible as i mentioned above. The limitation is no modifications can be made to those composite values such as taking only the first 2 letters of each composite value.

The way I see it in terms of the enhancement would look something like the following:
```js
// plugins.js
module.exports = ({ env }) => ({
  // ...
  slugify: {
    enabled: true,
    config: {
      contentTypes: {
        article: {
          field: 'slug',
          references: ['name','title'],
          transformer(record, referenceFields) {
            const { title, name } = referenceFields;
            return {
               title: title.substring(0,3),
               name: name.substring(0,3),
            }
          }
        },
      },
    },
  },
  // ...
});
```

In terms of adding a number at the end you can either add the entity/record id as a reference field (enhancement to be added in #28) or add it in the transformer function yourself and append it to an already existing field.

**Reply by 2-man** [reply]
> 

how to get only year from strapi dateformat? substring didnt work

**Reply by ComfortablyCoding** [reply]
This has not been implemented yet.

**Reply by zerenxyz** [reply]
> Just generating the composite UID is already possible via an array is already possible as i mentioned above. The limitation is no modifications can be made to those composite values such as taking only the first 2 letters of each composite value.
> 
> The way I see it in terms of the enhancement would look something like the following:
> 
> ```js
> // plugins.js
> module.exports = ({ env }) => ({
>   // ...
>   slugify: {
>     enabled: true,
>     config: {
>       contentTypes: {
>         article: {
>           field: 'slug',
>           references: ['name','title'],
>           transformer(record, referenceFields) {
>             const { title, name } = referenceFields;
>             return {
>                title: title.substring(0,3),
>                name: name.substring(0,3),
>             }
>           }
>         },
>       },
>     },
>   },
>   // ...
> });
> ```
> 
> In terms of adding a number at the end you can either add the entity/record id as a reference field (enhancement to be added in #28) or add it in the transformer function yourself and append it to an already existing field.

Havent had the chance and time to reply to this properly! Anyone then can just write their own return params I assume with this? Seems like a good fix, do you know any ETA when you would add this?

**Reply by ComfortablyCoding** [reply]
No ETA atm, need to find some time to implement and test this. PRs are always welcome.
